### PR TITLE
Provision Services as requested by the service graph

### PR DESCRIPTION
**Pull request by keeferrourke**
`CoordinatedService2` now takes a `Provider<Service>` instead of a `Service` in its constructor. 

This allows for the services to be provisioned as requested by the coordinated boot-up order in the service graph.

### CODE REVIEW

File: misk/src/main/kotlin/misk/CoordinatedService2.kt [hunk 1]
```diff
@@ -55,7 +56,14 @@ import com.google.common.util.concurrent.Service.State
  * This Service will stall in the `STARTING` state until all upstream services are `RUNNING`.
  * Symmetrically it stalls in the `STOPPING` state until all dependent services are `TERMINATED`.
  */
-internal class CoordinatedService2(val service: Service) : AbstractService() {
+internal class CoordinatedService2(
+    private val serviceProvider : Provider<out Service>
+) : AbstractService() {
+
+  val service : Service by lazy {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/CoordinatedService2.kt]: COOL

File: misk/src/main/kotlin/misk/MiskServiceModule.kt [hunk 2]
```diff
@@ -183,7 +182,7 @@ class MiskCommonServiceModule : KAbstractModule() {
 
 internal data class DependencyEdge(val dependent: Key<*>, val dependsOn: Key<*>)
 internal data class EnhancementEdge(val toBeEnhanced: Key<*>, val enhancement: Key<*>)
-internal data class ServiceEntry(val key: Key<out Service>)
+internal data class ServiceEntry(val key: Key<out Service>, val provider: Provider<out Service>)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/MiskServiceModule.kt]: omit the `Provider` field and compute it on-demand?

Comment (swankjesse) [reply misk/src/main/kotlin/misk/MiskServiceModule.kt]: (that’ll be `injector.getProvider(key)` on line 133)

Comment (keeferrourke) [reply misk/src/main/kotlin/misk/MiskServiceModule.kt]: fixed

**Pull request merged**